File: rust/plugins/network/ios/Sources/NetworkPlugin.swift
import Foundation
import Network
import SwiftRs
import Tauri
import WebKit

// Connectivity status via NWPathMonitor. `connected` means the path is
// satisfied (usable internet), `type` is the primary interface, and
// `metered` maps to the path being expensive (cellular / personal hotspot).
class NetworkPlugin: Plugin {
    private let monitor = NWPathMonitor()
    private var lastEmitted: String?

    @objc public override func load(webview: WKWebView) {
        super.load(webview: webview)
        // Watch the default path for the whole app lifetime; the monitor is
        // cheap and keeps nb:network-changed flowing with no JS setup. The
        // handler dedupes, since the monitor can re-report the same path.
        monitor.pathUpdateHandler = { [weak self] path in
            guard let self = self else { return }
            let (connected, type, metered) = Self.read(path)
            let key = "\(connected)|\(type)|\(metered)"
            if key == self.lastEmitted { return }
            self.lastEmitted = key

            var data = JSObject()
            data["connected"] = connected
            data["type"] = type
            data["metered"] = metered
            self.trigger("network-changed", data: data)
        }
        monitor.start(queue: DispatchQueue(label: "nativeblade-network"))
    }

    @objc public func getStatus(_ invoke: Invoke) {
        let (connected, type, metered) = Self.read(monitor.currentPath)
        invoke.resolve([
            "connected": connected,
            "type": type,
            "metered": metered,
        ])
    }

    private static func read(_ path: NWPath) -> (Bool, String, Bool) {
        let connected = path.status == .satisfied
        let type: String
        if !connected {
            type = "none"
        } else if path.usesInterfaceType(.wifi) {
            type = "wifi"
        } else if path.usesInterfaceType(.cellular) {
            type = "cellular"
        } else if path.usesInterfaceType(.wiredEthernet) {
            type = "ethernet"
        } else {
            type = "unknown"
        }
        return (connected, type, path.isExpensive)
    }
}

@_cdecl("init_plugin_nativeblade_network")
func initPlugin() -> Plugin {
    return NetworkPlugin()
}

File: rust/plugins/network/android/src/main/java/app/nativeblade/network/NetworkPlugin.kt
package app.nativeblade.network

import android.app.Activity
import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.webkit.WebView
import app.tauri.annotation.Command
import app.tauri.annotation.TauriPlugin
import app.tauri.plugin.Invoke
import app.tauri.plugin.JSObject
import app.tauri.plugin.Plugin

@TauriPlugin
class NetworkPlugin(private val activity: Activity) : Plugin(activity) {

    private val connectivity: ConnectivityManager
        get() = activity.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager

    // Last status forwarded to JS. Android delivers callback bursts (several
    // onCapabilitiesChanged per network switch), so only real changes emit.
    @Volatile
    private var lastEmitted: String? = null

    override fun load(webView: WebView) {
        super.load(webView)
        // Watch the default network for the whole app lifetime; the callback
        // is cheap and keeps nb:network-changed flowing with no JS setup.
        connectivity.registerDefaultNetworkCallback(object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network: Network) = emitIfChanged()
            override fun onLost(network: Network) = emitIfChanged()
            override fun onCapabilitiesChanged(network: Network, caps: NetworkCapabilities) =
                emitIfChanged()
        })
    }

    @Command
    fun getStatus(invoke: Invoke) {
        invoke.resolve(currentStatus())
    }

    private fun emitIfChanged() {
        val status = currentStatus()
        val key = status.toString()
        if (key == lastEmitted) return
        lastEmitted = key
        trigger("network-changed", status)
    }

    private fun currentStatus(): JSObject {
        val caps = connectivity.activeNetwork?.let { connectivity.getNetworkCapabilities(it) }
        // VALIDATED means actual internet, not just an interface that is up
        // (or a captive portal pretending to be one).
        val connected = caps?.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED) == true
        val type = when {
            caps == null -> "none"
            caps.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "wifi"
            caps.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "cellular"
            caps.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> "ethernet"
            else -> "unknown"
        }

        val obj = JSObject()
        obj.put("connected", connected)
        obj.put("type", type)
        obj.put("metered", connectivity.isActiveNetworkMetered)
        return obj
    }
}
